plugins/node/instrumentation-kafkajs/src/instrumentation.ts

```diff
@@ -0,0 +1,452 @@
+            config,
+            'eachBatch',
+            instrumentation._getConsumerEachBatchPatch()
+          );
+        }
+        return original.call(this, config);
+      };
+    };
+  }
+
+  private _getConsumerEachMessagePatch() {
+    const instrumentation = this;
+    return (original: ConsumerRunConfig['eachMessage']) => {
+      return function eachMessage(
+        this: unknown,
+        ...args: Parameters<EachMessageHandler>
+      ): Promise<void> {
+        const payload = args[0];
+        const propagatedContext: Context = propagation.extract(
+          ROOT_CONTEXT,
+          payload.message.headers,
+          bufferTextMapGetter
+        );
+        const span = instrumentation._startConsumerSpan(
+          payload.topic,
+          payload.message,
+          MESSAGINGOPERATIONVALUES_PROCESS,
+          propagatedContext
+        );
+
+        const eachMessagePromise = context.with(
+          trace.setSpan(propagatedContext, span),
+          () => {
+            return original!.apply(this, args);
+          }
+        );
+        return instrumentation._endSpansOnPromise([span], eachMessagePromise);
+      };
+    };
+  }
+
+  private _getConsumerEachBatchPatch() {
+    return (original: ConsumerRunConfig['eachBatch']) => {
+      const instrumentation = this;
+      return function eachBatch(
+        this: unknown,
+        ...args: Parameters<EachBatchHandler>
+      ): Promise<void> {
+        const payload = args[0];
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#topic-with-multiple-consumers
+        const receivingSpan = instrumentation._startConsumerSpan(
+          payload.batch.topic,
+          undefined,
+          MESSAGINGOPERATIONVALUES_RECEIVE,
+          ROOT_CONTEXT
+        );
+        return context.with(
+          trace.setSpan(context.active(), receivingSpan),
+          () => {
+            const spans = payload.batch.messages.map(
+              (message: KafkaMessage) => {
+                const propagatedContext: Context = propagation.extract(
+                  ROOT_CONTEXT,
+                  message.headers,
+                  bufferTextMapGetter
+                );
+                const spanContext = trace
+                  .getSpan(propagatedContext)
+                  ?.spanContext();
+                let origSpanLink: Link | undefined;
+                if (spanContext) {
+                  origSpanLink = {
+                    context: spanContext,
+                  };
+                }
+                return instrumentation._startConsumerSpan(
+                  payload.batch.topic,
+                  message,
+                  MESSAGINGOPERATIONVALUES_PROCESS,
+                  undefined,
+                  origSpanLink
+                );
+              }
+            );
+            const batchMessagePromise: Promise<void> = original!.apply(
+              this,
```

Comment:
Isn't this inherited from the parent scope (in _this_ case the `this` argument of `function eachBatch(this: unknown, ...args)`)?

Reply:
you're right as per MDN docs at the end of section https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#cannot_be_used_as_methods says

> For similar reasons, the [call()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call), [apply()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply), and [bind()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind) methods are not useful when called on arrow functions, **because arrow functions establish `this` based on the scope the arrow function is defined within, and the `this` value does not change based on how the function is invoked.**